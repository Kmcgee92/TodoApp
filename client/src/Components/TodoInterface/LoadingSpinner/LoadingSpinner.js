import React, {useState, useEffect} from "react";
import clsx from "clsx";
import CircularProgress from "@material-ui/core/CircularProgress";
import Fab from "@material-ui/core/Fab";
import CheckIcon from "@material-ui/icons/Check";
import CachedIcon from "@material-ui/icons/Cached";

export const LoadingSpinnerUX = ({classes, loading, setLoading, success, setSuccess}) => {

  const timer = React.useRef();

  useEffect(() => {
    setSuccess(false);
    setLoading(true);
    timer.current = window.setTimeout(() => {
      setSuccess(true);
      setLoading(false);
    }, 2000);
    return () => {
      clearTimeout(timer.current);
    };
  }, [setLoading, setSuccess]);

    const buttonClassname = clsx({
      [classes.buttonSuccess]: success,
      [classes.dataLoading]: !success,
    });
    
  return (
    <>
      <div className={classes.spinner}>
        <Fab aria-label="save" className={buttonClassname}>
          {success ? <CheckIcon /> : <CachedIcon />}
        </Fab>
        {loading && (
          <CircularProgress size={68} className={classes.fabProgress} />
        )}
      </div>
      <div className={classes.spinner}></div>
    </>
  );

}
const LoadingSpinner = ({ classes }) => {
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState(false);

  let props = {
    loading,
    setLoading,
    success,
    setSuccess,
    classes
  }
  return (
    <LoadingSpinnerUX {...props}/>
  );
};

export default LoadingSpinner;
