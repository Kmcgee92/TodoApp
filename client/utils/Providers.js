import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
// provide redux and apollo
import { Provider } from "react-redux";
import { MockedProvider } from "@apollo/client/testing";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

export default function ComponentWithProviders(
  ComponentToBeTested,
  props,
  initialReduxMocks,
  initialApolloMocks
) {
  let mocks = initialApolloMocks || [];
  let store = mockStore(
    
    initialReduxMocks || { auth: { activeUser: {}, error: false } }
  
  );
  return (
    <MockedProvider mocks={mocks}>
      <Provider store={store}>
        <ComponentToBeTested {...props} />
      </Provider>
    </MockedProvider>
  );
}
