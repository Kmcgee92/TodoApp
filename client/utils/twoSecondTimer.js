export default function twoSecondTmer() {
  return new Promise((resolve) => {
    setTimeout(resolve, 2000);
  });
}
