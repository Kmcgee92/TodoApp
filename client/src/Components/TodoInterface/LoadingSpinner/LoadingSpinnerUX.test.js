import React from "react";
import { mount, shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { act } from "react-dom/test-utils";
// component to test
import { LoadingSpinnerUX } from "./LoadingSpinner";
// mui component
import CircularProgress from "@material-ui/core/CircularProgress";
import CachedIcon from "@material-ui/icons/Cached";
import CheckIcon from "@material-ui/icons/Check";

// utils
import { findByTestAtrr } from "../../../../utils/index";
import twoSecondTmer from "../../../../utils/twoSecondTimer";

describe("<LoadingSpinner />", () => {
  let loadingSpinnerUX;
  let mockProps = {
    classes: {},
    loading: false,
    setLoading: jest.fn(),
    success: false,
    setSuccess: jest.fn(),
  };
  describe("Static State Testing", () => {
    it("Should Render", () => {
      loadingSpinnerUX = shallow(<LoadingSpinnerUX {...mockProps} />);
      expect(loadingSpinnerUX).toHaveLength(1);
      expect(toJSON(loadingSpinnerUX)).toMatchSnapshot();
    });
    it("Should display <CircularProgress/> and <CachedIcon/> on Static State Mount", () => {
      mockProps = {
        classes: {},
        loading: true,
        setLoading: jest.fn(),
        success: false,
        setSuccess: jest.fn(),
      };
      loadingSpinnerUX = shallow(<LoadingSpinnerUX {...mockProps} />);
      expect(loadingSpinnerUX.find(CircularProgress)).toHaveLength(1);
      expect(loadingSpinnerUX.find(CachedIcon)).toHaveLength(1);
    });
  });
  describe("State Change Testing", () => {
    it("Should not display <CircularProgress/> After State Change", () => {
      mockProps = {
        classes: {},
        loading: false,
        setLoading: jest.fn(),
        success: true,
        setSuccess: jest.fn(),
      };
      loadingSpinnerUX = shallow(<LoadingSpinnerUX {...mockProps} />);
      expect(loadingSpinnerUX.find(CircularProgress)).toHaveLength(0);
      expect(loadingSpinnerUX.find(CachedIcon)).toHaveLength(0);
      expect(loadingSpinnerUX.find(CheckIcon)).toHaveLength(1);
    });
    it("Should Trigger React useEffect on Mount and Execute Timeout", async () => {
      loadingSpinnerUX = mount(<LoadingSpinnerUX {...mockProps} />);
      expect(mockProps.setLoading).toBeCalledTimes(1);
      expect(mockProps.setSuccess).toBeCalledTimes(1);
      await twoSecondTmer();
      expect(mockProps.setLoading).toBeCalledTimes(2);
      expect(mockProps.setSuccess).toBeCalledTimes(2);
    });
  });
});
