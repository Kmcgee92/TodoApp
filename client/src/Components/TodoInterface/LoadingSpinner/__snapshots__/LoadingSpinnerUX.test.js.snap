// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingSpinner /> Static State Testing Should Render 1`] = `
<Fragment>
  <div>
    <WithStyles(ForwardRef(Fab))
      aria-label="save"
      className="undefined"
    >
      <Memo([object Object]) />
    </WithStyles(ForwardRef(Fab))>
  </div>
  <div />
</Fragment>
`;
